File: packages/lyrics/swift/Sources/BridgethingLyrics/LrclibResolver.swift
import Foundation

#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

/// `LyricsResolver` backed by lrclib.net's signature lookup endpoint.
/// No auth, free, community-uploaded LRC. Coverage is strong on western
/// mainstream and weak on remixes / live / regional / very new releases.
public final class LrclibResolver: LyricsResolver, @unchecked Sendable {
    public let name: String = "lrclib"

    private let baseURL: URL
    private let userAgent: String
    private let session: URLSession

    public init(
        baseURL: URL = URL(string: "https://lrclib.net")!,
        userAgent: String = "bridgething/0.1 (+https://github.com/thinglabsoss/bridgething)",
        session: URLSession = .shared
    ) {
        self.baseURL = baseURL
        self.userAgent = userAgent
        self.session = session
    }

    public func lyrics(for track: TrackIdentity) async -> Lyrics? {
        guard let url = makeQueryURL(for: track) else { return nil }

        var request = URLRequest(url: url)
        request.setValue(userAgent, forHTTPHeaderField: "User-Agent")
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
                return nil
            }
            let entry = try JSONDecoder().decode(LrclibEntry.self, from: data)
            return entry.toLyrics()
        } catch {
            return nil
        }
    }

    private func makeQueryURL(for track: TrackIdentity) -> URL? {
        var components = URLComponents(url: baseURL.appendingPathComponent("api/get"), resolvingAgainstBaseURL: false)
        var items: [URLQueryItem] = [
            URLQueryItem(name: "artist_name", value: track.artist),
            URLQueryItem(name: "track_name", value: track.track),
        ]
        if let album = track.album {
            items.append(URLQueryItem(name: "album_name", value: album))
        }
        if let durationMs = track.durationMs {
            items.append(URLQueryItem(name: "duration", value: String(durationMs / 1000)))
        }
        components?.queryItems = items
        return components?.url
    }
}

private struct LrclibEntry: Decodable {
    let plainLyrics: String?
    let syncedLyrics: String?
    let instrumental: Bool?

    func toLyrics() -> Lyrics? {
        if instrumental == true {
            return Lyrics(synced: nil, plain: nil, source: "lrclib")
        }
        let synced = syncedLyrics.flatMap { LRCParser.parse($0) }
        if synced == nil, (plainLyrics ?? "").isEmpty {
            return nil
        }
        return Lyrics(
            synced: synced?.isEmpty == true ? nil : synced,
            plain: plainLyrics?.isEmpty == true ? nil : plainLyrics,
            source: "lrclib"
        )
    }
}

/// Parses the LRC time-tagged format. Each line may carry one or more
/// `[mm:ss.xx]` timestamps followed by the line text; lines without any
/// timestamp are ignored. Multiple timestamps on a single line emit
/// multiple `LyricLine` entries with the same text.
enum LRCParser {
    static func parse(_ text: String) -> [LyricLine] {
        var out: [LyricLine] = []
        for rawLine in text.split(separator: "\n", omittingEmptySubsequences: false) {
            let line = String(rawLine)
            let (timestamps, body) = extractTimestamps(line)
            if timestamps.isEmpty { continue }
            for ms in timestamps {
                out.append(LyricLine(startMs: ms, text: body))
            }
        }
        return out.sorted(by: { $0.startMs < $1.startMs })
    }

    private static func extractTimestamps(_ line: String) -> (timestamps: [Int], body: String) {
        var stamps: [Int] = []
        var rest = Substring(line)
        while rest.hasPrefix("[") {
            guard let close = rest.firstIndex(of: "]") else { break }
            let inside = rest[rest.index(after: rest.startIndex)..<close]
            if let ms = parseTimestamp(String(inside)) {
                stamps.append(ms)
                rest = rest[rest.index(after: close)...]
            } else {
                break
            }
        }
        let body = String(rest).trimmingCharacters(in: .whitespaces)
        return (stamps, body)
    }

    private static func parseTimestamp(_ s: String) -> Int? {
        let parts = s.split(separator: ":")
        guard parts.count == 2 else { return nil }
        guard let minutes = Int(parts[0]) else { return nil }
        let secondParts = parts[1].split(separator: ".")
        guard let seconds = Int(secondParts[0]) else { return nil }
        var hundredths = 0
        if secondParts.count == 2 {
            let frac = secondParts[1]
            let normalized: String
            if frac.count == 2 {
                normalized = String(frac)
            } else if frac.count == 3 {
                normalized = String(frac.prefix(2))
            } else {
                normalized = String(frac).padding(toLength: 2, withPad: "0", startingAt: 0)
            }
            hundredths = Int(normalized) ?? 0
        }
        return (minutes * 60 + seconds) * 1000 + hundredths * 10
    }
}

File: packages/lyrics/kotlin/lyrics/src/main/kotlin/dev/bridgething/lyrics/LrclibResolver.kt
package dev.bridgething.lyrics

import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.engine.cio.CIO
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.client.request.parameter
import io.ktor.client.statement.HttpResponse
import io.ktor.http.HttpStatusCode
import io.ktor.serialization.kotlinx.json.json
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * `LyricsResolver` backed by lrclib.net's signature lookup endpoint.
 * No auth, free, community-uploaded LRC. Coverage is strong on western
 * mainstream and weak on remixes / live / regional / very new releases.
 */
class LrclibResolver(
    private val baseUrl: String = "https://lrclib.net",
    private val userAgent: String = "bridgething/0.1 (+https://github.com/thinglabsoss/bridgething)",
    httpClient: HttpClient? = null,
) : LyricsResolver {
    override val name: String = "lrclib"

    private val client: HttpClient = httpClient ?: HttpClient(CIO) {
        install(ContentNegotiation) {
            json(Json { ignoreUnknownKeys = true })
        }
    }

    override suspend fun lyrics(track: TrackIdentity): Lyrics? {
        val response: HttpResponse = try {
            client.get("$baseUrl/api/get") {
                header("User-Agent", userAgent)
                header("Accept", "application/json")
                parameter("artist_name", track.artist)
                parameter("track_name", track.track)
                track.album?.let { parameter("album_name", it) }
                track.durationMs?.let { parameter("duration", (it / 1000).toString()) }
            }
        } catch (e: Exception) {
            return null
        }

        if (response.status != HttpStatusCode.OK) return null

        val entry: LrclibEntry = try {
            response.body()
        } catch (e: Exception) {
            return null
        }

        return entry.toLyrics()
    }
}

@Serializable
private data class LrclibEntry(
    val plainLyrics: String? = null,
    val syncedLyrics: String? = null,
    val instrumental: Boolean? = null,
) {
    fun toLyrics(): Lyrics? {
        if (instrumental == true) {
            return Lyrics(synced = null, plain = null, source = "lrclib")
        }
        val synced = syncedLyrics?.let { LRCParser.parse(it) }?.takeIf { it.isNotEmpty() }
        val plain = plainLyrics?.takeIf { it.isNotEmpty() }
        if (synced == null && plain == null) return null
        return Lyrics(synced = synced, plain = plain, source = "lrclib")
    }
}

/**
 * Parses the LRC time-tagged format. Each line may carry one or more
 * `[mm:ss.xx]` timestamps followed by the line text; lines without any
 * timestamp are ignored. Multiple timestamps on a single line emit
 * multiple `LyricLine` entries with the same text.
 */
object LRCParser {
    fun parse(text: String): List<LyricLine> {
        val out = mutableListOf<LyricLine>()
        for (rawLine in text.split('\n')) {
            val (timestamps, body) = extractTimestamps(rawLine)
            if (timestamps.isEmpty()) continue
            for (ms in timestamps) {
                out += LyricLine(startMs = ms, text = body)
            }
        }
        return out.sortedBy { it.startMs }
    }

    private fun extractTimestamps(line: String): Pair<List<Int>, String> {
        val stamps = mutableListOf<Int>()
        var rest = line
        while (rest.startsWith("[")) {
            val close = rest.indexOf(']')
            if (close < 0) break
            val inside = rest.substring(1, close)
            val ms = parseTimestamp(inside) ?: break
            stamps += ms
            rest = rest.substring(close + 1)
        }
        return stamps to rest.trim()
    }

    private fun parseTimestamp(s: String): Int? {
        val parts = s.split(':')
        if (parts.size != 2) return null
        val minutes = parts[0].toIntOrNull() ?: return null
        val secondParts = parts[1].split('.')
        val seconds = secondParts[0].toIntOrNull() ?: return null
        var hundredths = 0
        if (secondParts.size == 2) {
            val frac = secondParts[1]
            val normalized = when (frac.length) {
                2 -> frac
                3 -> frac.substring(0, 2)
                else -> frac.padEnd(2, '0').substring(0, 2)
            }
            hundredths = normalized.toIntOrNull() ?: 0
        }
        return (minutes * 60 + seconds) * 1000 + hundredths * 10
    }
}
